core/templates/pages/exploration-editor-page/services/graph-data.service.ts

```diff
@@ -16,46 +16,49 @@
-  'ComputeGraphService', 'ExplorationInitStateNameService',
-  'ExplorationStatesService',
-  function(
-      ComputeGraphService, ExplorationInitStateNameService,
-      ExplorationStatesService) {
-    var _graphData = null;
-    var _recomputeGraphData = function() {
-      if (!ExplorationInitStateNameService.savedMemento) {
-        return;
-      }
-
-      var states = ExplorationStatesService.getStates();
-      var initStateId = ExplorationInitStateNameService.savedMemento;
-      _graphData = ComputeGraphService.compute(initStateId, states);
-    };
-
-    return {
-      recompute: function() {
-        _recomputeGraphData();
-      },
-      /**
-       * @return graphData - Directed graph visualization input. This object
-       * includes the following keys:
-       * - nodes: Objects with keys of nodeids and values of node names.
-       * - links: list of objects. Each object represents a directed link
-       *    between two nodes, and has keys 'source' and 'target', the values
-       *    of which are the names of the corresponding nodes.
-       * - initStateName: the name of the initial state.
-       * - finalStateName: the name of the final state.
-       */
-      getGraphData: function() {
-        return angular.copy(_graphData);
-      }
-    };
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { ComputeGraphService, GraphData } from 'services/compute-graph.service';
+import { ExplorationInitStateNameService } from 'pages/exploration-editor-page/services/exploration-init-state-name.service';
+import { ExplorationStatesService } from 'pages/exploration-editor-page/services/exploration-states.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class GraphDataService {
+  _graphData: GraphData | null = null;
+
+  constructor(
+    private computeGraphService: ComputeGraphService,
+    private explorationInitStateNameService: ExplorationInitStateNameService,
+    private explorationStatesService: ExplorationStatesService
+  ) {}
+
+  recompute(): void {
+    if (!this.explorationInitStateNameService.savedMemento) {
+      return;
+    }
+
+    var states = this.explorationStatesService.getStates();
+    var initStateId = this.explorationInitStateNameService.savedMemento;
+    this._graphData = this.computeGraphService.compute(initStateId, states);
+  }
+
+  /**
+   * @return graphData - Directed graph visualization input. This object
+   * includes the following keys:
+   * - nodes: Objects with keys of nodeids and values of node names.
+   * - links: list of objects. Each object represents a directed link
+   *    between two nodes, and has keys 'source' and 'target', the values
+   *    of which are the names of the corresponding nodes.
+   * - initStateName: the name of the initial state.
+   * - finalStateName: the name of the final state.
+   */
+  getGraphData(): GraphData {
+    return angular.copy(this._graphData);
```

Comment:
Due to same reasons as mentioned above, prefer `_.cloneDeep` here.

Reply:
Done.